// @materials-ui/core/dist/esm/materials-dropdown.entry.js
import { r as registerInstance, h } from './index-c0dc768b.js';

const Dropdown = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
    }
    componentDidLoad() {
        this.materialsMenuEl.setPosition(this.menuPosition);
        if (this.buttonType)
            this.buttonEl[this.buttonType] = true;
    }
    componentWillUpdate() {
        if (this.buttonType)
            this.buttonEl[this.buttonType] = true;
    }
    async openMenu() {
        this.materialsMenuEl.open();
    }
    async closeMenu() {
        this.materialsMenuEl.close();
    }
    async isOpen() {
        return this.materialsMenuEl.isOpen();
    }
    renderIconHelper() {
        return this.iconHelper ? h("materials-icon", { style: { 'margin-left': '8px', 'margin-right': '-8px' }, name: this.iconHelper }) : '';
    }
    render() {
        return (h("div", null, this.iconOnly &&
            h("materials-icon-button", { onClick: () => this.openMenu(), icon: this.icon }), !this.iconOnly &&
            h("materials-button", { color: this.color, onClick: () => this.openMenu(), icon: this.icon, ref: el => this.buttonEl = el }, this.menuTitle, this.renderIconHelper()), h("materials-menu", { ref: el => this.materialsMenuEl = el }, h("slot", null))));
    }
    static get style() { return "materials-list-item>a{color:inherit;text-decoration:none}materials-menu{white-space:nowrap}"; }
};

export { Dropdown as materials_dropdown };
