// @materials-ui/core/dist/esm/materials-datatable-pagination.entry.js
import { r as registerInstance, c as createEvent, h } from './index-c0dc768b.js';

const DatatablePagination = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.size = 0;
        this.total = 0;
        this.currentPage = 0;
        this.first = 0;
        this.last = 0;
        this.nextPage = createEvent(this, "nextPage", 7);
        this.previousPage = createEvent(this, "previousPage", 7);
    }
    handleNextClick() {
        this.nextPage.emit();
    }
    handlePreviousClick() {
        this.previousPage.emit();
    }
    render() {
        return (h("div", { class: "container" }, h("materials-layout-grid", { class: "pagination" }, h("materials-layout-cell", { class: "navigate" }, h("materials-icon-button", { icon: "navigate_before", onClick: () => this.handlePreviousClick() })), h("materials-layout-cell", { class: "pages" }, h("span", null, " ", this.first, " - ", this.last, " sur ", this.total)), h("materials-layout-cell", { class: "navigate" }, h("materials-icon-button", { icon: "navigate_next", onClick: () => this.handleNextClick() })))));
    }
    static get style() { return ".container{float:right}.container,.pagination{display:-ms-inline-flexbox;display:inline-flex}.pagination{color:var(--mdc-theme-on-surface);opacity:.8}.pagination>div{padding:0}.pages{padding-top:12px}.pages>span{display:block;width:150px}.navigate,.pages>span{text-align:center}"; }
};

export { DatatablePagination as materials_datatable_pagination };
