// @materials-ui/core/dist/esm/materials-datatable-header.entry.js
import { r as registerInstance, c as createEvent, h } from './index-c0dc768b.js';

const DatatableHeader = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        this.selectAll = createEvent(this, "selectAll", 7);
    }
    watchSelected() {
        this.selectAll.emit(this.selected);
    }
    /**
     * Définir le checkbox du header
     */
    renderSelectBox() {
        if (this.selectable) {
            return h("materials-checkbox", { color: "primary", onChange: () => {
                    this.selected = !this.selected;
                }, onClick: (e) => e.stopPropagation(), checked: this.selected });
        }
    }
    getClasses() {
        return {
            'container': true,
            'container-background-primary': this.color === 'primary',
            'container-background-secondary': this.color === 'secondary',
        };
    }
    render() {
        return (h("div", { class: this.getClasses() }, this.renderSelectBox(), h("slot", null)));
    }
    static get watchers() { return {
        "selected": ["watchSelected"]
    }; }
    static get style() { return ":host{display:block}:host .container{display:-ms-flexbox;display:flex;-ms-flex-direction:row;flex-direction:row;-ms-flex-align:center;align-items:center;border-bottom:1px solid #e0e0e0;background-color:var(--mdc-theme-datatable-header,#fff);color:var(--mdc-theme-on-datatable-header,#212121);padding:0 8px}:host materials-checkbox{-ms-flex-item-align:center;align-self:center}"; }
};

export { DatatableHeader as materials_datatable_header };
