// @materials-ui/core/dist/cjs/materials-datatable-pagination.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-c3bf379a.js');

const DatatablePagination = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.size = 0;
        this.total = 0;
        this.currentPage = 0;
        this.first = 0;
        this.last = 0;
        this.nextPage = index.createEvent(this, "nextPage", 7);
        this.previousPage = index.createEvent(this, "previousPage", 7);
    }
    handleNextClick() {
        this.nextPage.emit();
    }
    handlePreviousClick() {
        this.previousPage.emit();
    }
    render() {
        return (index.h("div", { class: "container" }, index.h("materials-layout-grid", { class: "pagination" }, index.h("materials-layout-cell", { class: "navigate" }, index.h("materials-icon-button", { icon: "navigate_before", onClick: () => this.handlePreviousClick() })), index.h("materials-layout-cell", { class: "pages" }, index.h("span", null, " ", this.first, " - ", this.last, " sur ", this.total)), index.h("materials-layout-cell", { class: "navigate" }, index.h("materials-icon-button", { icon: "navigate_next", onClick: () => this.handleNextClick() })))));
    }
    static get style() { return ".container{float:right}.container,.pagination{display:-ms-inline-flexbox;display:inline-flex}.pagination{color:var(--mdc-theme-on-surface);opacity:.8}.pagination>div{padding:0}.pages{padding-top:12px}.pages>span{display:block;width:150px}.navigate,.pages>span{text-align:center}"; }
};

exports.materials_datatable_pagination = DatatablePagination;
